tensorflow/python/estimator/canned/boosted_trees.py

```diff
@@ -910,8 +939,103 @@ def _create_regression_head(label_dimension, weight_column=None):
+  importances = np.zeros(num_features)
+
+  for node in tree.nodes:
+    node_type = node.WhichOneof('node')
+    if node_type == 'bucketized_split':
+      feature_id = node.bucketized_split.feature_id
+      importances[feature_id] += node.metadata.gain
+    elif node_type == 'leaf':
+      assert node.metadata.gain == 0
+    else:
+      raise ValueError('Unexpected split type %s', node_type)
+
+  return importances
+
+
+def _compute_feature_importances(tree_ensemble, num_features, normalize):
+  """Compute the feature importances.
+
+  The higher the value, the more important the feature.
+
+  Args:
+    tree_ensemble: a trained tree ensemble, instance of proto
+      boosted_trees.TreeEnsemble.
+    num_features: The total number of feature ids.
+    normalize: If True, normalize the feature importances.
+
+  Returns:
+    sorted_feature_idx: A list of feature_id which is sorted
+      by its feature importance.
+    feature_importances: A list of corresponding feature importances.
+
+  Raises:
+    AssertionError: If feature importances contain negative value.
+      Or if normalize = True and normalization is not possible
+      (e.g. ensemble is empty or trees contain only a root node).
+  """
+  tree_importances = [_compute_feature_importances_per_tree(tree, num_features)
+                      for tree in tree_ensemble.trees]
+  tree_importances = np.array(tree_importances)
+  tree_weights = np.array(tree_ensemble.tree_weights).reshape(-1, 1)
+  feature_importances = np.sum(tree_importances * tree_weights, axis=0)
+  assert np.all(feature_importances >= 0), ('feature_importances '
+                                            'must be non-negative.')
+  if normalize:
+    normalizer = np.sum(feature_importances)
+    assert normalizer > 0, 'Trees are all empty or contains only a root node.'
+    feature_importances /= normalizer
+
+  sorted_feature_idx = np.argsort(feature_importances)[::-1]
+  return sorted_feature_idx, feature_importances[sorted_feature_idx]
+
+
+class _BoostedTrees(estimator.Estimator):
+  """Base class for boosted trees estimators."""
+
+  def __init__(self, model_fn, model_dir, config, feature_columns):
+    super(_BoostedTrees, self).__init__(
+        model_fn=model_fn, model_dir=model_dir, config=config)
+
+    self._sorted_feature_columns = sorted(feature_columns, key=lambda tc: tc.name)
+    self._num_features = _calculate_num_features(self._sorted_feature_columns)
+
+  def experimental_feature_importances(self, normalize=False):
+    """Computes gain-based feature importances.
+
+    The higher the value, the more important the corresponding feature.
+
+    Args:
+      normalize: If True, normalize the feature importances.
+
+    Returns:
+      sorted_feature_names: 1-D array of feature name which is sorted
+        by its feature importance.
+      feature_importances: 1-D array of the corresponding feature importance.
+
+    Raises:
+      ValueError: Empty ensemble.
+    """
```

Comment:
When attempting to normalize on an empty ensemble or an ensemble of trees which have no splits?